2003-08-29  Gwenole Beauchesne  <gbeauchesne@mandrakesoft.com>

	* sysdeps/i386/dl-machine.h (dl_platform_init): Avoid i686
	libraries to be loaded if the host doesn't support CMOV.

--- glibc-2.3.2/sysdeps/i386/dl-machine.h.i586-if-no-cmov	2003-07-07 17:15:04.000000000 +0200
+++ glibc-2.3.2/sysdeps/i386/dl-machine.h	2003-08-29 10:37:21.000000000 +0200
@@ -347,6 +347,9 @@ _dl_start_user:\n\
    Prelinked libraries may use Elf32_Rela though.  */
 #define ELF_MACHINE_PLT_REL 1
 
+/* We need the definition of HWCAP_I386_CMOV.  */
+#include <sysdeps/unix/sysv/linux/i386/dl-hwcapinfo.h>
+
 /* We define an initialization functions.  This is called very early in
    _dl_sysdep_start.  */
 #define DL_PLATFORM_INIT dl_platform_init ()
@@ -357,6 +357,12 @@ dl_platform_init (void)
   if (GL(dl_platform) != NULL && *GL(dl_platform) == '\0')
     /* Avoid an empty string which would disturb us.  */
     GL(dl_platform) = NULL;
+
+  if (GL(dl_platform) != NULL
+      && GL(dl_platform)[1] >= '6' && (GL(dl_hwcap) & HWCAP_I386_CMOV) == 0)
+    /* Avoid i686 compiled libraries to be loaded if the host doesn't
+       support CMOV instructions.  */
+    GL(dl_platform) = "i586";
 }
 
 static inline Elf32_Addr
